astropy/modeling/bounding_box.py

```diff
@@ -0,0 +1,578 @@
+        Determine which parts of an input array are outside the interval.
+
+    domain :
+        Contructs a discretization of the points inside the interval.
+    """
+
+    @staticmethod
+    def _validate_shape(interval):
+        """Validate the shape of an interval representation"""
+        MESSAGE = """An interval must be some sort of sequence of length 2"""
+
+        try:
+            shape = np.shape(interval)
+        except TypeError:
+            try:
+                # np.shape does not work with lists of Quantities
+                if len(interval) == 1:
+                    interval = interval[0]
+                shape = np.shape([b.to_value() for b in interval])
+            except (ValueError, TypeError, AttributeError):
+                raise ValueError(MESSAGE)
+
+        valid_shape = shape in ((2,), (1, 2), (2, 0))
+        if not valid_shape:
+            valid_shape = (len(shape) > 0) and (shape[0] == 2) and \
+                all(isinstance(b, np.ndarray) for b in interval)
+
+        if not isiterable(interval) or not valid_shape:
+            raise ValueError(MESSAGE)
+
+    @classmethod
+    def _validate_bounds(cls, lower, upper):
+        """Validate the bounds are reasonable and construct an interval from them."""
+        if (np.asanyarray(lower) > np.asanyarray(upper)).all():
+            warnings.warn(f"Invalid interval: upper bound {upper} "
+                          f"is strictly less than lower bound {lower}.", RuntimeWarning)
+
+        return cls(lower, upper)
+
+    @classmethod
+    def validate(cls, interval):
+        """
+        Construct and validate an interval
+
+        Parameters
+        ----------
+        interval : iterable
+            A representation of the interval.
+
+        Returns
+        -------
+        A validated interval.
+        """
+        cls._validate_shape(interval)
+
+        if len(interval) == 1:
+            interval = tuple(interval[0])
+        else:
+            interval = tuple(interval)
+
+        return cls._validate_bounds(interval[0], interval[1])
+
+    def outside(self, _input: np.ndarray):
+        """
+        Parameters
+        ----------
+        _input : np.ndarray
+            The evaluation input in the form of an array.
+
+        Returns
+        -------
+        Boolean array indicating which parts of _input are outside the interval:
+            True  -> position outside interval
+            False -> position inside  interval
+        """
+        return np.logical_or(_input < self.lower, _input > self.upper)
+
+    def domain(self, resolution):
+        return np.arange(self.lower, self.upper + resolution, resolution)
+
+
+class BoundingBox(object):
+    """
+    A model's bounding box
+
+    Parameters
+    ----------
+    intervals : dict
+        A dictionary containing all the intervals for each model input
+            keys   -> input index
+            values -> interval for that index
+
+    _model : astropy.modeling.Model
```

Comment:
I don't understand why the leading underscore is needed here. This parameter is exposed in a public API but leading underscore implies that user should not touch it.

Reply:
The current bounding box uses the prefixed underscore in its interfaces, see https://github.com/astropy/astropy/blob/0b81dfdf0fe2bf41ec15a4d505d84ce53143adf6/astropy/modeling/utils.py#L430-L452.

Reply:
Sure, but `_BoundingBox` was "hidden"... Hmm... I guess I see the point of not breaking the names but you are already renaming everything else... 🤷  Furthermore, it was hidden before, so there is nothing stopping you from removing the leading underscore now.

Reply:
I made the changes in c180ff9.